File: Haskell/src/Jogo/MenuControlador.hs
module Jogo.MenuControlador (iniciarMenu) where

import qualified UI.Menu as Menu
import qualified UI.UtilsUI as UtilsUI
import qualified UI.HUD as HUD
 
import Modelos.Jogo
import Modelos.Bot
import Modelos.Mercado (comprarItem)
import Modelos.Tabuleiro as Tabuleiro
import Modelos.Jogador as Jogador

import Jogo.GameOver as GameOver

import System.Random(randomRIO)
import System.Console.ANSI (clearScreen)
import Data.Char(isDigit, toUpper)

-- =========================================================================MENU================================================================================

iniciarMenu :: IO ()
iniciarMenu = do
  clearScreen 
  Menu.logoMenu
  opcaoEnter <- getLine
  clearScreen
  Menu.opcoesMenu
  opcao <- getLine
  processarOpcao opcao

processarOpcao :: String -> IO ()
processarOpcao "1" = do
  clearScreen
  Menu.novoJogo
  nomeJogador <- getLine
  iniciarJogo nomeJogador
processarOpcao "2" = do
  clearScreen
  Menu.carregarJogo
  slot <- getLine
  case slot of 
    _ | slot `elem` ["1", "2", "3"] -> carregarJogo slot
    _ | slot `elem` ["V", "v"] -> subMenu
    _ -> processarOpcao "2"
processarOpcao "3" = do
  clearScreen
  Menu.historiaJogo
  processarSubOpcao
processarOpcao "4" = do
  clearScreen
  Menu.regrasJogo
  processarSubOpcao
processarOpcao "5" = do
  clearScreen
  Menu.creditosJogo
  processarSubOpcao
processarOpcao "6" = do
    putStr UtilsUI.confirmacao
    opcao <- getLine
    case opcao of 
        op | op `elem` ["S", "s"] -> do
            clearScreen
            putStrLn "Saindo do jogo..."
        _ -> subMenu
processarOpcao _ = do
  putStr UtilsUI.opcaoInvalida
  opcao <- getLine
  processarOpcao opcao
  
-- Funções auxiliares MENU

subMenu :: IO ()
subMenu = do
    clearScreen
    Menu.opcoesMenu
    opcao <- getLine
    processarOpcao opcao

processarSubOpcao :: IO ()
processarSubOpcao = do
    opcao <- getLine
    case opcao of
        op | op `elem` ["V", "v"] -> do
            clearScreen
            subMenu
        _ -> do
            putStr UtilsUI.voltarMenu
            processarSubOpcao

-- =========================================================================JOGO================================================================================

iniciarJogo :: String -> IO ()
iniciarJogo nomeJogador = do
    jogo <- inicializarJogo nomeJogador  
    loopJogo jogo  

carregarJogo :: FilePath -> IO ()
carregarJogo numero = do
    let caminho = "src/BD/save" ++ numero ++ ".json"
    jogoCarregado <- carregarSave caminho
    case jogoCarregado of
        Just jogo -> loopJogo jogo
        Nothing   -> putStrLn "Erro: Não foi possível carregar o jogo." 

loopJogo :: Jogo -> IO ()
loopJogo jogo = do
    clearScreen
    HUD.mainScreen jogo
    gameOver <- GameOver.verificaVitoriaDerrotaPlayer jogo
    if gameOver
      then do 
        processarSubOpcao
      else do
        opcao <- getLine
        processarOpcaoLoop opcao jogo

processarOpcaoLoop :: String -> Jogo -> IO ()
processarOpcaoLoop "1" jogo = do
    let tabelaPosDrone = desfazVisualizacaoDrone (tabela (jogador jogo))
    if tabelaPosDrone /= tabela (jogador jogo)
      then 
        processarOpcaoLoop "1" (jogo { jogador = (jogador jogo) { tabela = tabelaPosDrone } })
      else
        if getBombasPequenas (getJogador jogo) <= 0
          then 
            loopJogo jogo
          else do
              (coluna, linha) <- inputCoordenada
              let (novaTabelaJog, novasMoedas) = 
                      atirouNaCoordenada (tabela (jogador jogo)) 
                          (getIndexColuna ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L"] coluna 0) 
                          (linha - 1)
              r <- randomRIO(0,143)
              let jogoAtualizado = jogo { jogador = (jogador jogo) 
                  { tabela = novaTabelaJog
                  , bombasPequenas = bombasPequenas (jogador jogo) - 1  
                  , moedas = getMoedas (jogador jogo) + novasMoedas  
                  }, bot = (jogar (bot jogo) r)
              }
              loopJogo jogoAtualizado

processarOpcaoLoop "2" jogo = do
  let tabelaPosDrone = desfazVisualizacaoDrone (tabela (jogador jogo))
  if tabelaPosDrone /= tabela (jogador jogo)
    then 
      processarOpcaoLoop "2" (jogo { jogador = (jogador jogo) { tabela = tabelaPosDrone } })
    else
      if getBombasMedias (getJogador jogo) <= 0
        then 
          loopJogo jogo
        else do
            (coluna, linha) <- inputCoordenada
            let (novaTabelaJog, novasMoedas) = 
                    tiroBombaMedia (tabela (jogador jogo)) 
                        (getIndexColuna ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L"] coluna 0) 
                        (linha - 1)
            r <- randomRIO(0,143)
            let jogoAtualizado = jogo { jogador = (jogador jogo) 
                { tabela = novaTabelaJog
                , bombasMedias = bombasMedias (jogador jogo) - 1  
                , moedas = getMoedas (jogador jogo) + novasMoedas  
                }, bot = (jogar (bot jogo) r)
            }
            loopJogo jogoAtualizado

processarOpcaoLoop "3" jogo = do
    let tabelaPosDrone = desfazVisualizacaoDrone (tabela (jogador jogo))
    if tabelaPosDrone /= tabela (jogador jogo)
      then 
        processarOpcaoLoop "3" (jogo { jogador = (jogador jogo) { tabela = tabelaPosDrone } })
      else
        if getBombasGrandes (getJogador jogo) <= 0
          then 
            loopJogo jogo
          else do
              (coluna, linha) <- inputCoordenada
              let (novaTabelaJog, novasMoedas) = 
                      tiroBombaGrande (tabela (jogador jogo)) 
                          (getIndexColuna ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L"] coluna 0) 
                          (linha - 1)
              r <- randomRIO(0,143)
              let jogoAtualizado = jogo { jogador = (jogador jogo) 
                  { tabela = novaTabelaJog
                  , bombasGrandes = bombasGrandes (jogador jogo) - 1  
                  , moedas = getMoedas (jogador jogo) + novasMoedas  
                  }, bot = (jogar (bot jogo) r)
              }
              loopJogo jogoAtualizado

processarOpcaoLoop "4" jogo = do
  if getDroneVisualizador (getJogador jogo) <= 0 then loopJogo jogo
  else do
    (coluna, linha) <- inputCoordenada
    let novaTabelaJog = 
          drone (tabela (jogador jogo)) 
            (getIndexColuna ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L"] coluna 0) 
            (linha - 1)
    let jogoAtualizado = jogo { jogador = (jogador jogo) { tabela = novaTabelaJog, droneVisualizador = droneVisualizador (jogador jogo) - 1 } }
    loopJogo jogoAtualizado

processarOpcaoLoop "m" jogo = do
  clearScreen
  HUD.mercadoScreen jogo
  item <- getLine
  if item `elem` ["1", "2", "3"]
    then do
      let jogador = getJogador jogo
      let mercado = getMercado jogo
      let novoJogador = Modelos.Mercado.comprarItem jogador mercado item
      let novoJogo = setJogador novoJogador jogo  
      loopJogo novoJogo
    else processarSubOpcaoLoop jogo item

processarOpcaoLoop "s" jogo = do
  clearScreen
  HUD.saveJogoScreen
  slot <- getLine
  let caminho = "src/BD/save" ++ slot ++ ".json"
  if slot `elem` ["1", "2", "3"]
    then do
      putStr UtilsUI.confirmacao
      confirmacaoSalvamento caminho jogo slot
    else processarSubOpcaoLoop jogo slot
processarOpcaoLoop "q" jogo = do
  putStr UtilsUI.confirmacao
  opcao <- getLine
  case opcao of 
      op | op `elem` ["S", "s"] -> subMenu
      _ -> do
        putStr "> Digite a opção: "
        opcao <- getLine
        processarOpcaoLoop opcao jogo

processarOpcaoLoop _ jogo = do
  putStr UtilsUI.opcaoInvalida
  opcao <- getLine
  processarOpcaoLoop opcao jogo

-- Funções auxiliares JOGO

processarSubOpcaoLoop :: Jogo -> String -> IO ()
processarSubOpcaoLoop jogo opcao
  | opcao `elem` ["V", "v"] = loopJogo jogo
  | otherwise = do
                putStr UtilsUI.opcaoInvalidaVoltar
                opcaoNovamente <- getLine
                processarSubOpcaoLoop jogo opcaoNovamente 

confirmacaoSalvamento :: FilePath -> Jogo -> String -> IO ()
confirmacaoSalvamento caminho jogo slot = do
  opcao <- getLine
  case opcao of  
    op | op `elem` ["S", "s"] -> do
      clearScreen
      Modelos.Jogo.salvarJogo caminho jogo
      carregarJogo slot  
    _ -> do
      loopJogo jogo

getIndexColuna :: [String] -> String -> Int -> Int 
getIndexColuna listaLetras str i =
  if listaLetras !! i == str 
    then i
    else getIndexColuna listaLetras str (i + 1)

inputCoordenada :: IO (String, Int)
inputCoordenada = do
    putStr "> Digite a coordenada: "
    coord <- getLine
    if validarCoordenada coord
        then let (coluna:linha) = coord
             in return ([toUpper coluna], read linha)
        else do
            putStrLn UtilsUI.opcaoInvalida
            inputCoordenada

validarCoordenada :: String -> Bool
validarCoordenada (c:ls) =
    c `elem` ['A'..'L'] && all isDigit ls && let l = read ls in l `elem` [1..12]
validarCoordenada _ = False  

File: Prolog/src/Jogo/MenuControlador.pl
:- module(menu_controlador, [iniciarMenu/0]).

:- use_module('../UI/HUD').
:- use_module('../UI/Menu').
:- use_module('../UI/UtilsUI').

:- use_module('../Modelos/Jogo').
:- use_module('../Modelos/Jogador').
:- use_module('../Modelos/Mercado').
:- use_module('../Modelos/Tabuleiro').
:- use_module('./GameOver').

:- use_module(library(lists)).
:- use_module(library(charsio)).
:- use_module(library(random)).

% Limpar a tela
clear_screen :- 
    write('\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n').

% =========================================================================MENU================================================================================

iniciarMenu :-
    clear_screen,
    logoMenu,
    read_line_to_string(user_input, _),
    clear_screen,
    opcoesMenu,
    read_line_to_string(user_input, Opcao),
    processarOpcao(Opcao).

processarOpcao("1") :-
    clear_screen,
    novoJogo,
    read_line_to_string(user_input, NomeJogador),
    iniciarJogo(NomeJogador). 

processarOpcao("2") :-
    clear_screen,
    carregarJogo,
    read_line_to_string(user_input, Slot),
    (   member(Slot, ["1", "2","3"]) -> carregarJogo(Slot)
    ;   member(Slot, ["V", "v"]) -> subMenu
    ;   processarOpcao("2")).

processarOpcao("3") :-
    clear_screen,
    historiaJogo,
    processarSubOpcao.

processarOpcao("4") :-
    clear_screen,
    regrasJogo,
    processarSubOpcao.

processarOpcao("5") :-
    clear_screen,
    creditosJogo,
    processarSubOpcao.

processarOpcao("6") :-
    confirmacao,
    read_line_to_string(user_input, Opcao),
    (   member(Opcao, ["S", "s"]) -> clear_screen, writeln('Saindo do jogo...'), halt
    ;   subMenu).

processarOpcao(_) :-
    opcaoInvalida,
    read_line_to_string(user_input, Opcao),
    processarOpcao(Opcao).

% Regras auxiliares para voltar ao menu principal ou subopções dentro do menu
subMenu :-
    clear_screen,
    opcoesMenu,
    read_line_to_string(user_input, Opcao),
    processarOpcao(Opcao).

processarSubOpcao :-
    read_line_to_string(user_input, Opcao),
    (   member(Opcao, ["V", "v"]) -> clear_screen, subMenu
    ;   voltarMenu, processarSubOpcao).

% =========================================================================JOGO================================================================================

iniciarJogo(NomeJogador) :-
    inicializarJogo(NomeJogador, Jogo),
    loopJogo(Jogo).

carregarJogo(Numero) :-
    string_concat('../BD/save', Numero, Caminho),
    carregarSave(Caminho, Resultado),
    loopJogo(Resultado).
carregarJogo(_) :-
    writeln('Erro: Não foi possível carregar o jogo.'),
    iniciarMenu.

loopJogo(Jogo) :- 
    clear_screen,
    mainScreen(Jogo),
    verificaVitoriaDerrotaPlayer(Jogo, GameOver),
    (
        GameOver == true ->
            processarSubOpcao
        ;
            read_line_to_string(user_input, Opcao),
            processarOpcaoLoop(Opcao, Jogo)
    ).

processarOpcaoLoop("1", Jogo) :-
    getJogador(Jogo, Jogador),
    getTabelaJogador(Jogador, TabelaAtual),
    desfazVisualizacaoDrone(TabelaAtual, TabelaSemDrone),
    (
        TabelaSemDrone \= TabelaAtual ->
            setTabelaJogador(Jogador, TabelaSemDrone, NovoJogador),
            setJogador(Jogo, NovoJogador, NovoJogo),
            processarOpcaoLoop("1", NovoJogo)
    ;
        getBombasPequenas(Jogador, BP),
        BP =< 0 ->
            loopJogo(Jogo)
    ;
        inputCoordenada(ColunaStr, Linha),
        getIndexColuna(['A','B','C','D','E','F','G','H','I','J','K','L'], ColunaStr, 0, ColunaIndex),
        LinhaIndex is Linha - 1,
        mainAtirouNaCoordenada(TabelaAtual, LinhaIndex, ColunaIndex, NovaTabelaJog, NovasMoedas),

        random_between(0, 143, R),
        getMoedas(Jogador, MoedasAtuais),
        NovoTotalMoedas is MoedasAtuais + NovasMoedas,
        getBombasPequenas(Jogador, BP),
        NovoNumBombas is BP - 1,

        setTabelaJogador(Jogador, NovaTabelaJog, JogadorTemp1),
        setBombasPequenas(JogadorTemp1, NovoNumBombas, JogadorTemp2),
        setMoedas(JogadorTemp2, NovoTotalMoedas, NovoJogador),

        % Aqui o bot deveria jogar
        getBot(Jogo, Bot),
        % jogarBot(Bot, R, NovoBot),

        getMercado(Jogo, Mercado),
        getDataJogo(Jogo, DataJogo),
        JogoAtualizado = jogo(NovoJogador, Bot, Mercado, DataJogo),
        loopJogo(JogoAtualizado)
    ).

processarOpcaoLoop("2", Jogo) :-
    getJogador(Jogo, Jogador),
    getTabelaJogador(Jogador, TabelaAtual),
    desfazVisualizacaoDrone(TabelaAtual, TabelaSemDrone),
    (
        TabelaSemDrone \= TabelaAtual ->
            setTabelaJogador(Jogador, TabelaSemDrone, NovoJogador),
            setJogador(Jogo, NovoJogador, NovoJogo),
            processarOpcaoLoop("2", NovoJogo)
    ;
        getBombasMedias(Jogador, BM),
        BM =< 0 ->
            loopJogo(Jogo)
    ;
        inputCoordenada(ColunaStr, Linha),
        getIndexColuna(['A','B','C','D','E','F','G','H','I','J','K','L'], ColunaStr, 0, ColunaIndex),
        LinhaIndex is Linha - 1,
        tiroBombaMedia(TabelaAtual, LinhaIndex, ColunaIndex, NovaTabelaJog, NovasMoedas),

        random_between(0, 143, R),
        getMoedas(Jogador, MoedasAtuais),
        NovoTotalMoedas is MoedasAtuais + NovasMoedas,
        getBombasMedias(Jogador, BM),
        NovoNumBombas is BM - 1,

        setTabelaJogador(Jogador, NovaTabelaJog, JogadorTemp1),
        setBombasMedias(JogadorTemp1, NovoNumBombas, JogadorTemp2),
        setMoedas(JogadorTemp2, NovoTotalMoedas, NovoJogador),

        % Aqui o bot deveria jogar
        getBot(Jogo, Bot),
        % jogarBot(Bot, R, NovoBot),

        getMercado(Jogo, Mercado),
        getDataJogo(Jogo, DataJogo),
        JogoAtualizado = jogo(NovoJogador, Bot, Mercado, DataJogo),
        loopJogo(JogoAtualizado)
    ).

processarOpcaoLoop("3", Jogo) :-
    getJogador(Jogo, Jogador),
    getTabelaJogador(Jogador, TabelaAtual),
    desfazVisualizacaoDrone(TabelaAtual, TabelaSemDrone),
    (
        TabelaSemDrone \= TabelaAtual ->
            setTabelaJogador(Jogador, TabelaSemDrone, NovoJogador),
            setJogador(Jogo, NovoJogador, NovoJogo),
            processarOpcaoLoop("3", NovoJogo)
    ;
        getBombasGrandes(Jogador, BG),
        BG =< 0 ->
            loopJogo(Jogo)
    ;
        inputCoordenada(ColunaStr, Linha),
        getIndexColuna(['A','B','C','D','E','F','G','H','I','J','K','L'], ColunaStr, 0, ColunaIndex),
        LinhaIndex is Linha - 1,
        tiroBombaGrande(TabelaAtual, LinhaIndex, ColunaIndex, NovaTabelaJog, NovasMoedas),

        random_between(0, 143, R),
        getMoedas(Jogador, MoedasAtuais),
        NovoTotalMoedas is MoedasAtuais + NovasMoedas,
        getBombasGrandes(Jogador, BG),
        NovoNumBombas is BG - 1,

        setTabelaJogador(Jogador, NovaTabelaJog, JogadorTemp1),
        setBombasGrandes(JogadorTemp1, NovoNumBombas, JogadorTemp2),
        setMoedas(JogadorTemp2, NovoTotalMoedas, NovoJogador),

        % Aqui o bot deveria jogar
        getBot(Jogo, Bot),
        % jogarBot(Bot, R, NovoBot),

        getMercado(Jogo, Mercado),
        getDataJogo(Jogo, DataJogo),
        JogoAtualizado = jogo(NovoJogador, NovoBot, Mercado, DataJogo),
        loopJogo(JogoAtualizado)
    ).

processarOpcaoLoop("4", Jogo) :-
    getJogador(Jogo, Jogador),
    getDroneVisualizador(Jogador, NumDrones),
    (
        NumDrones =< 0 ->
            loopJogo(Jogo)
    ;
        inputCoordenada(ColunaStr, Linha),
        getTabelaJogador(Jogador, TabelaAtual),
        getIndexColuna(['A','B','C','D','E','F','G','H','I','J','K','L'], ColunaStr, 0, ColunaIndex),
        LinhaIndex is Linha - 1,

        drone(TabelaAtual, ColunaIndex, LinhaIndex, NovaTabelaJog),

        NovoNumDrones is NumDrones - 1,
        setTabelaJogador(Jogador, NovaTabelaJog, JogadorTemp),
        setDroneVisualizador(JogadorTemp, NovoNumDrones, NovoJogador),

        getBot(Jogo, Bot),
        getMercado(Jogo, Mercado),
        getDataJogo(Jogo, DataJogo),
        JogoAtualizado = jogo(NovoJogador, Bot, Mercado, DataJogo),
        loopJogo(JogoAtualizado)
    ).  

processarOpcaoLoop("m", Jogo) :- 
    clear_screen,
    mercadoScreen(Jogo),
    read_line_to_string(user_input, Item),
    ( member(Item, ["1", "2", "3"]) ->
        getJogador(Jogo, Jogador),
        getMercado(Jogo, Mercado),
        comprarItem(Jogador, Mercado, Item, NovoJogador),
        setJogador(Jogo, NovoJogador, NovoJogo),
        loopJogo(NovoJogo)
    ; processarSubOpcaoLoop(Jogo, Item)
    ).

processarOpcaoLoop("s", Jogo) :-
    clear_screen,
    saveJogoScreen,
    read_line_to_string(user_input, Slot),
    atom_concat("../BD/save", Slot, Caminho),
    ( member(Slot, ["1", "2", "3"]) ->
        confirmacao,
        confirmacaoSalvamento(Caminho, Jogo, Slot)
    ; processarSubOpcaoLoop(Jogo, Slot)
    ).

processarOpcaoLoop("q", Jogo) :- 
    confirmacao,
    read_line_to_string(user_input, Opcao),
    ( member(Opcao, ["S", "s"]) ->
        subMenu
    ; write("> Digite a opção: "),
      read_line_to_string(user_input, NovaOpcao),
      processarOpcaoLoop(NovaOpcao, Jogo)
    ).

processarOpcaoLoop(_, Jogo) :-
    opcaoInvalida,
    read_line_to_string(user_input, Opcao),
    processarOpcaoLoop(Opcao, Jogo).

% Funções auxiliares JOGO

processarSubOpcaoLoop(Jogo, Opcao) :-
    member(Opcao, ["V", "v"]),
    loopJogo(Jogo).

processarSubOpcaoLoop(Jogo, _) :-
    opcaoInvalidaVoltar,
    read_line_to_string(user_input, NovaOpcao),
    processarSubOpcaoLoop(Jogo, NovaOpcao).

confirmacaoSalvamento(Caminho, Jogo, Slot) :-
    read_line_to_string(user_input, Opcao),
    ( member(Opcao, ["S", "s"]) ->
        clear_screen,
        salvarJogo(Caminho, Jogo),
        carregarJogo(Slot)
    ; loopJogo(Jogo)
    ).

getIndexColuna([H|_], H, Index, Index).
getIndexColuna([_|T], Elem, CurrIndex, Index) :-
    NextIndex is CurrIndex + 1,
    getIndexColuna(T, Elem, NextIndex, Index).

inputCoordenada(ColunaChar, Numero) :-
    write('> Digite a coordenada (ex: A5): '),
    read_line_to_string(user_input, Input),
    (
        validarCoordenada(Input) ->
            string_chars(Input, [LetraChar | Digitos]),
            ColunaChar = LetraChar, 
            atom_chars(AtomNumero, Digitos),
            atom_number(AtomNumero, Numero)
        ;
            opcaoInvalida, nl,
            inputCoordenada(ColunaChar, Numero)
    ).

validarCoordenada(Coordenada) :-
    string_chars(Coordenada, [LetraChar | Digitos]),
    char_type(LetraChar, upper),
    member(LetraChar, ['A','B','C','D','E','F','G','H','I','J','K','L']),
    Digitos \= [],
    maplist(char_type_digit, Digitos),
    atom_chars(AtomNumero, Digitos),
    atom_number(AtomNumero, NumLinha),
    NumLinha >= 1,
    NumLinha =< 12.

char_type_digit(C) :- char_type(C, digit).